Choose between FrugalFeeds and Couponese (issue #204)

tj-dodd opened the pull request:
This gives the user the ability to choose between FrugalFeeds and Couponese through a source optional argument. I ended up making it so the bot's default behaviour is to give the user a randomly generated combined list of coupons. I changed the hyperlink for the embed to FrugalFeeds, as there's a more comprehensive collection of coupon codes there.

49Indium commented on uqcsbot/dominos_coupons.py:
```diff
@@ -139,27 +157,88 @@ def _get_coupons(n: int, ignore_expiry: bool, keywords: List[str]) -> List[Coupo
         ]
-    return coupons[:n]
 
+    if not source:
```
This is a great idea. Given the coupons are never ordered, the coupons should be shuffled regardless of `source`.

49Indium commented on uqcsbot/dominos_coupons.py:
```diff
@@ -139,27 +157,88 @@ def _get_coupons(n: int, ignore_expiry: bool, keywords: List[str]) -> List[Coupo
+            code_container = soup_coupon.find(class_=siteclass.get("code"))
+
+            if (
+                not expiry_date_container
```
Potentially we might be able to still show coupons that don't have a date, but this is up to you.

tj-dodd replied:
I'll have a look into this, however the user is generally provided enough coupon codes where we can get away with excluding coupons that don't have a date. (and I'm not sure if I've come across a coupon code without a date yet)

49Indium commented on uqcsbot/dominos_coupons.py:
```diff
@@ -139,27 +157,88 @@ def _get_coupons(n: int, ignore_expiry: bool, keywords: List[str]) -> List[Coupo
+                    # Convert shortened month to numerical value
+                    month: int = datetime.strptime(date_values[1], "%b").month
+                except (ValueError, IndexError):
+                    continue
```
Even if the coupon has an unidentifiable month, it should probably still be included (maybe with question marks for the month).

tj-dodd replied:
I agree with this - I admit this was mostly a band-aid solution at the time as I was dealing with invalid date formats causing index errors.

49Indium commented on uqcsbot/dominos_coupons.py:
```diff
@@ -139,27 +157,88 @@ def _get_coupons(n: int, ignore_expiry: bool, keywords: List[str]) -> List[Coupo
+            description: str = description_container.get_text(strip=True)
+            code: str = code_container.get_text(strip=True)
+
+            if url == FRUGAL_FEEDS_DOMINOS_URL:
```
This one is up to you, but it might be nice to restructure the `Coupon` class to have the date as an `Optional[Datetime]` instead of a string (or maybe even a `Datetime|str`, where the Datetime is used by default, but if it can't be parsed, the `str` is used to still provide some information about the date). This would allow us to have all date parsing here, regardless of URL used.

49Indium reviewed (changes requested):
Firstly, thank you for contributing! This is a great improvement and should help reduce the trial and error needed when getting pizza!

Overall the command adjustments appear to to be working great. I like how you have structured `_get_coupons_from_page` and the idea of shuffling coupons each time. There are some small adjustments I think would be nice which I've attached with the code. Some of these comments are up to your discretion; they are all relatively minor and could be argued either way if they are needed.

tj-dodd commented on uqcsbot/dominos_coupons.py:
```diff
@@ -54,39 +58,51 @@ async def dominoscoupons(
+            request_url = error.request.url if error.request else "Unknown site."
             logging.warning(
-                f"Could not connect to dominos coupon site ({COUPONESE_DOMINOS_URL}): {resp_content}"
+                f"Could not connect to dominos coupon site ({request_url}): {resp_content}"
```
This is a great idea, I'll start on working on getting this implemented.

tj-dodd commented:
Just committed addressing your changes - if there's any issues let me know.

I ended up reverting back to how you were originally parsing dates. My additions were unnecessary I think - to be transparent, I'm not completely sure what I was trying to do, but I believe at one point I mistakenly assumed I had to do some extra parsing as the month formatting on FrugalFeeds is slightly different and I was encountering a couple of weird dates (e.g. soon!). I think leaving the date as a string if it's unable to be parsed is a much better implementation. There's really only 1-2 ambiguous dates in the FrugalFeeds list and the format is slightly different to Couponese, but the user still receives more than enough information even if the date is passed through as an unformatted string.

I'm still not certain if giving users a random sample of discount codes is the best way to approach things, as people might get unlucky (e.g. bot giving them 5 different mudcake discounts, excellent!), but I guess it's on the user to make use of the keywords function, and if you're getting decent results then I guess it's fine.

tj-dodd commented:
nevermind, I think most recent commit is the best way to do the formatting

tj-dodd commented:
Some small changes:

- Bot now hints user to change source if their keywords don't return any results on a single website
- FrugalFeeds went rogue and started advertising pizza hut deals on their dominos page, these have been filtered out
- Fixed FrugalFeeds formatting for discounts with multiple codes
- Moved hyperlinks down to description area, bot now links both FrugalFeeds and Couponese if both are selected

49Indium commented on uqcsbot/dominos_coupons.py:
```diff
@@ -44,50 +48,105 @@ async def dominoscoupons(
             return
 
+        if source == "both":
+            description_string = f"Sourced from [FrugalFeeds]({COUPONESE_DOMINOS_URL}) and [Couponese]({FRUGAL_FEEDS_DOMINOS_URL})"
```
This has the links the wrong way around

tj-dodd replied:
Not my proudest moment

49Indium commented on uqcsbot/dominos_coupons.py:
```diff
@@ -44,50 +48,105 @@ async def dominoscoupons(
             )
             return
+
+        if source != "both":
```
This shouldn't be needed any more, given that the source can only be "both", "couponese" or "frugalfeeds" (and should pass type checking).

49Indium commented on uqcsbot/dominos_coupons.py:
```diff
@@ -44,50 +48,105 @@ async def dominoscoupons(
+        coupons: List[Coupon] = []
+        stored_url: str = " "
+
+        for tries in range(0, NUMBER_WEBSITES):
```
Instead of looping over a range, it might be better to loop over a list of the URLs you are going to gather coupons from. You appear to do this within `_get_coupons_from_page`,  so I'm unsure why multiple attempts are needed here.

tj-dodd replied:
I really like this idea (I admit my implementation could be better), as well as the idea you left below in the unresolved comment. I'll begin working on this, and the code should be a lot cleaner afterwards.

49Indium commented on uqcsbot/dominos_coupons.py:
```diff
@@ -44,50 +48,105 @@ async def dominoscoupons(
+            await interaction.edit_original_response(content=content_str)
             return
 
+        if source == "both":
```
I like the way that you have included this in the embed.

49Indium commented on uqcsbot/dominos_coupons.py:
```diff
@@ -123,12 +182,14 @@ def keyword_matches(self, keyword: str) -> bool:
     """
 
-    coupons = _get_coupons_from_page()
+    coupons: List[Coupon] = _get_coupons_from_page(source)
```
The explicit type on this line isn't needed. Feel free to leave it if you think it makes it more readable.

49Indium commented on uqcsbot/dominos_coupons.py:
```diff
@@ -139,27 +200,95 @@ def _get_coupons(n: int, ignore_expiry: bool, keywords: List[str]) -> List[Coupo
 
 
-def _get_coupons_from_page() -> List[Coupon]:
+def _get_coupons_from_page(source: str) -> List[Coupon]:
```
It would be nice to use the literal type again for `source`. This should avoid some of your case checking below. Additionally, I'd personally make this function only accept a single source (so not "both"), and let `_get_coupons` turn multiple sources into multiple calls of this function.

49Indium commented on uqcsbot/dominos_coupons.py:
```diff
@@ -139,27 +200,95 @@ def _get_coupons(n: int, ignore_expiry: bool, keywords: List[str]) -> List[Coupo
+            code: str = code_container.get_text(strip=True)
+
+            temp_code: str = code.replace(",", "")
+            if not temp_code.isdigit():
```
I'm assuming this is to remove the frugalfeed coupons that aren't for dominos.
However, this appears to remove some couponese coupons (the ones that have spaces in them) as well.
I think a comment here explaining why this occurs would be good.

49Indium commented on uqcsbot/dominos_coupons.py:
```diff
@@ -44,50 +48,105 @@ async def dominoscoupons(
         number_of_coupons: int = 5,
         ignore_expiry: bool = True,
```
Unrelated to the changes you have made, but it might be best to make this `False` by default. This should provide a better list of coupons,  at the risk of having less coupons to print out.

tj-dodd replied:
Good call

49Indium commented on uqcsbot/dominos_coupons.py:
```diff
@@ -139,27 +200,95 @@ def _get_coupons(n: int, ignore_expiry: bool, keywords: List[str]) -> List[Coupo
+            if not temp_code.isdigit():
+                continue
+
+            code = code.replace(",", " ")
```
I think using commas to separate codes (rather than spaces) is better. This is just my personal opinion, so it's up to you.

tj-dodd replied:
Good call, I'll make it so they're comma separated - the bot originally had spaces separating the codes so I just assumed you wanted to stick with that style

49Indium replied:
I had (incorrectly) assumed that the codes with spaces were just longer codes, not two different codes. Feel more than free to make stylistic changes as you see fit.

49Indium reviewed (changes requested):
Thanks for making those changes. It's annoying that FrugalFeeds has included different codes.
I'm unsure how much I like the structure of the large for loop (iterating over `range(0, NUMBER_WEBSITES)` within the main command function. Looping over a list of urls in `_get_coupons` or `_get_coupons_from_page` might be a better alternative.

tj-dodd commented:
An update:

- Requested changes implemented
- Now checks for and removes duplicate codes
- A good amount of restructuring in `__get_coupons_from_page`, `_get_coupons`, and `DominosCoupons`

49Indium commented on uqcsbot/dominos_coupons.py:
```diff
@@ -183,17 +158,62 @@ def keyword_matches(self, keyword: str) -> bool:
+    n: int,
+    ignore_expiry: bool,
+    keywords: List[str],
+    source: Literal["couponese", "frugalfeeds", "both"],
```
Given that you use `Literal["couponese", "frugalfeeds", "both"]`, it might be best to create a new type (see [`advent.py`](https://github.com/UQComputingSociety/uqcsbot-discord/blob/main/uqcsbot/advent.py#L45) for an example).

tj-dodd replied:
I am fairly unfamiliar with creating new types, but I have had a crack at implementing this. I'm unsure if I'm following best practices here:

`CouponSource = Literal[
    "frugalfeeds", 
    "couponese", 
    "both"]`

I've declared this globally and have replaced all instances of `Literal["couponese", "frugalfeeds", "both"]` declaration with CouponSource, e.g. 

`source: CouponSource = "both"`, 
`def _get_coupons(
    n: int,
    ignore_expiry: bool,
    keywords: List[str],
    source: CouponSource,
) -> Tuple[List[Coupon], List[str]]:`,

And then have used `source` as I was using it before, e.g.. 

`coupons, failed_urls = _get_coupons(
            number_of_coupons, ignore_expiry, keywords.split(), source
        )`

`if source == "both":
            description_string = f"Sourced from [FrugalFeeds]({FRUGAL_FEEDS_DOMINOS_URL}) and [Couponese]({COUPONESE_DOMINOS_URL})"`

Am I on the right path here? Additionally, I spotted another possible area of improvement - I've used `Literal["couponese", "frugalfeeds"]]` a couple of times. Should I declare it as its own type as well? Is there a better way to approach this? It's just `CouponSource` without both.

49Indium replied:
That's exactly it. It makes the code easier to read and reduces repetition. The `SingleSource` type is also a goo idea. Personally, I would have made `CouponSource = SingleSource | Literal["both"]`, but how you have implemented it is also very clean IMO.

49Indium commented on uqcsbot/dominos_coupons.py:
```diff
@@ -183,17 +158,62 @@ def keyword_matches(self, keyword: str) -> bool:
     if not ignore_expiry:
         coupons = [coupon for coupon in coupons if coupon.is_valid()]
 
+    # Remove duplicates
```
This is a great idea

49Indium commented on uqcsbot/dominos_coupons.py:
```diff
@@ -203,15 +223,18 @@ def _get_coupons(
-    urls: List[str] = []
-    coupons: List[Coupon] = []
+
+    urls: Dict[str, str] = {
```
I'd potentially make this dictionary a global constant, but that's up to you.

tj-dodd replied:
Do you recommend a certain style? Is it fine to leave it as `urls` or should it be capitalized i.e. `URLS`? I'm aware constants should be fully capitalized but I'm unsure as it's a dictionary.

49Indium replied:
All capitalised would be my thoughts, but I'm not certain.

49Indium commented on uqcsbot/dominos_coupons.py:
```diff
@@ -226,69 +249,64 @@ def _get_coupons_from_page(source: str) -> List[Coupon]:
+        # Keep formatting same for coupons with multiple codes
+        if source == "frugalfeeds":
+            code = code.replace(",", ", ")
+        else:
```
Instead of using `else`, it might be best to use `if source == "couponese"`, as this could be forgotten if we add more sources in the future.

49Indium commented on uqcsbot/dominos_coupons.py:
```diff
@@ -226,69 +249,64 @@ def _get_coupons_from_page(source: str) -> List[Coupon]:
+        description: str = description_container.get_text(strip=True)
+        code: str = code_container.get_text(strip=True)
+
+        # Check if code is a digit - take seperators out so we can check
```
Why do we need to check if the code is a digit? Is this for the pizza hut coupons (that FrugalFeeds sometimes provides)? If it is, state that explicitly in your comment.

tj-dodd replied:
Yes, this is to deal with the pizza hut advertisements, I've since clarified this through a comment - FrugalFeeds have since taken down these advertisements but they might re-emerge in the future. They might also introduce some other form of advertising so I thought I should implement a somewhat robust filter, although there's probably a better way to do this so I'm open to any ideas.

49Indium replied:
I think this is good (especially with the new descriptive comment)

49Indium reviewed (changes requested):
This is getting close to ready. I like the duplicate checking! A few small changes would be nice to make the typing more readable and reduce repetition.

tj-dodd commented:
Changes done 👍
